nengo/builder/optimizer.py

```diff
@@ -0,0 +1,969 @@
+        replacements = {}
+
+        # Simple merge if all X are the same.
+        if all(o.X is ops[0].X for o in ops):
+            A = merge_signals_or_views([o.A for o in ops], replacements)
+            Y = merge_signals_or_views([o.Y for o in ops], replacements)
+            return operator.DotInc(A, ops[0].X, Y), replacements
+
+        # BSR merge if X differ
+        X = merge_signals_or_views([o.X for o in ops], replacements)
+        Y = merge_signals_or_views([o.Y for o in ops], replacements)
+
+        # Construct sparse A representation
+        data = np.array([o.A.initial_value for o in ops])
+        indptr = np.arange(len(ops) + 1, dtype=int)
+        indices = np.arange(len(ops), dtype=int)
+        name = 'bsr_merged<{first}, ..., {last}>'.format(
+            first=ops[0].A.name, last=ops[-1].A.name)
+        readonly = all([o.A.readonly for o in ops])
+        A = Signal(data, name=name, readonly=readonly)
+        for i, s in enumerate([o.A for o in ops]):
+            replacements[s] = Signal(
+                data[i], name="%s[%i]" % (s.name, i), base=A)
+            assert np.all(s.initial_value == replacements[s].initial_value)
+            assert s.shape == replacements[s].shape
+
+        reshape = operator.reshape_dot(
+            ops[0].A.initial_value, ops[0].X.initial_value,
+            ops[0].Y.initial_value, tag=ops[0].tag)
+        return (
+            operator.BsrDotInc(
+                A, X, Y, indices=indices, indptr=indptr, reshape=reshape),
+            replacements)
+
+
+@OpMergeOptimizer.register_merger(SimNeurons)
+class SimNeuronsMerger(AbstractMerger):
+    def can_merge(self, op1, op2):
+        return (
+            op1.neurons == op2.neurons and
+            all(compatible(s) for s in zip(op1.all_signals, op2.all_signals)))
+
+    def _gather(self, ops, key):
+        return [getattr(o, key) for o in ops]
+
+    def merge(self, ops):
+        replacements = {}
+        J = merge_signals_or_views(self._gather(ops, 'J'), replacements)
+        output = merge_signals_or_views(
+            self._gather(ops, 'output'), replacements)
+        states = []
+        for signals in zip(*self._gather(ops, 'states')):
+            states.append(merge_signals_or_views(signals, replacements))
+        return SimNeurons(ops[0].neurons, J, output, states), replacements
+
+
+def compatible(signals, axis=0):
+    """Checks that all signals have a compatible shape along a given axis
+    to allow for concatenation.
+
+    For views this includes also a check that the signals have a common
+    base and agree on the strides.
+
+    In comparison to the `check_*_mergeable` functions, this function does
+    not throw exceptions (either behavior might be desired dependent on the
+    context) and allows for either signals or signal views.
+    """
+    if len(set(signals)) != len(signals):
+        # Signal appears twice in list.
+        return False
+    for s in signals:
+        if s.ndim != signals[0].ndim:
```

Comment:
The main argument for `elif` seems to be in case of mutually exclusive conditions, it is less likely to introduce an error when someone changes the code. I feel in this case it less likely without the `elif` because the conditions are not mutually exclusive. If one of the `return` statements is removed, it should still return `False` if any of the other `if`-conditions evaluates to false.